File: Sources/FWObjC/Plugin/Refresh/RefreshPluginImpl.m
//
//  RefreshPluginImpl.m
//  FWFramework
//
//  Created by wuyong on 2022/8/23.
//

#import "RefreshPluginImpl.h"

#if FWMacroSPM

@interface UIScrollView ()

@property (nonatomic, strong, nullable) __FWPullRefreshView *__fw_pullRefreshView;
@property (nonatomic, assign) BOOL __fw_showPullRefresh;
@property (nonatomic, strong, nullable) __FWInfiniteScrollView *__fw_infiniteScrollView;
@property (nonatomic, assign) BOOL __fw_showInfiniteScroll;
@property (nonatomic, assign) BOOL __fw_infiniteScrollFinished;
- (void)__fw_addPullRefreshWithBlock:(void (^)(void))block;
- (void)__fw_addPullRefreshWithTarget:(id)target action:(SEL)action;
- (void)__fw_triggerPullRefresh;
- (void)__fw_addInfiniteScrollWithBlock:(void (^)(void))block;
- (void)__fw_addInfiniteScrollWithTarget:(id)target action:(SEL)action;
- (void)__fw_triggerInfiniteScroll;

@end

#else

#import <FWFramework/FWFramework-Swift.h>

#endif

#pragma mark - __FWRefreshPluginImpl

@implementation __FWRefreshPluginImpl

+ (__FWRefreshPluginImpl *)sharedInstance {
    static __FWRefreshPluginImpl *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[__FWRefreshPluginImpl alloc] init];
    });
    return instance;
}

#pragma mark - Refreshing

- (BOOL)isRefreshing:(UIScrollView *)scrollView {
    return scrollView.__fw_pullRefreshView.state == __FWPullRefreshStateLoading;
}

- (BOOL)shouldRefreshing:(UIScrollView *)scrollView {
    return scrollView.__fw_showPullRefresh;
}

- (void)setShouldRefreshing:(BOOL)shouldRefreshing scrollView:(UIScrollView *)scrollView {
    scrollView.__fw_showPullRefresh = shouldRefreshing;
}

- (void)setRefreshingBlock:(void (^)(void))block scrollView:(UIScrollView *)scrollView {
    [scrollView __fw_addPullRefreshWithBlock:block];
    if (self.pullRefreshBlock) self.pullRefreshBlock(scrollView.__fw_pullRefreshView);
}

- (void)setRefreshingTarget:(id)target action:(SEL)action scrollView:(UIScrollView *)scrollView {
    [scrollView __fw_addPullRefreshWithTarget:target action:action];
    if (self.pullRefreshBlock) self.pullRefreshBlock(scrollView.__fw_pullRefreshView);
}

- (void)beginRefreshing:(UIScrollView *)scrollView {
    [scrollView __fw_triggerPullRefresh];
}

- (void)endRefreshing:(UIScrollView *)scrollView {
    [scrollView.__fw_pullRefreshView stopAnimating];
}

#pragma mark - Loading

- (BOOL)isLoading:(UIScrollView *)scrollView {
    return scrollView.__fw_infiniteScrollView.state == __FWInfiniteScrollStateLoading;
}

- (BOOL)shouldLoading:(UIScrollView *)scrollView {
    return scrollView.__fw_showInfiniteScroll;
}

- (void)setShouldLoading:(BOOL)shouldLoading scrollView:(UIScrollView *)scrollView {
    scrollView.__fw_showInfiniteScroll = shouldLoading;
}

- (BOOL)loadingFinished:(UIScrollView *)scrollView {
    return scrollView.__fw_infiniteScrollFinished;
}

- (void)setLoadingFinished:(BOOL)loadingFinished scrollView:(UIScrollView *)scrollView {
    scrollView.__fw_infiniteScrollFinished = loadingFinished;
}

- (void)setLoadingBlock:(void (^)(void))block scrollView:(UIScrollView *)scrollView {
    [scrollView __fw_addInfiniteScrollWithBlock:block];
    if (self.infiniteScrollBlock) self.infiniteScrollBlock(scrollView.__fw_infiniteScrollView);
}

- (void)setLoadingTarget:(id)target action:(SEL)action scrollView:(UIScrollView *)scrollView {
    [scrollView __fw_addInfiniteScrollWithTarget:target action:action];
    if (self.infiniteScrollBlock) self.infiniteScrollBlock(scrollView.__fw_infiniteScrollView);
}

- (void)beginLoading:(UIScrollView *)scrollView {
    [scrollView __fw_triggerInfiniteScroll];
}

- (void)endLoading:(UIScrollView *)scrollView {
    [scrollView.__fw_infiniteScrollView stopAnimating];
}

@end

File: Sources/FWFramework/Plugin/Refresh/RefreshPluginImpl.swift
//
//  RefreshPluginImpl.swift
//  FWFramework
//
//  Created by wuyong on 2022/8/23.
//

import UIKit
#if FWMacroSPM
import FWObjC
#endif

/// 默认刷新插件
open class RefreshPluginImpl: NSObject, RefreshPlugin {
    
    // MARK: - Accessor
    /// 单例模式
    @objc(sharedInstance)
    public static let shared = RefreshPluginImpl()
    
    /// 上拉追加是否显示完成视图，默认true
    open var showsFinishedView = true
    
    /// 下拉刷新自定义句柄，开启时自动调用
    open var pullRefreshBlock: ((PullRefreshView) -> Void)?
    
    /// 上拉追加自定义句柄，开启时自动调用
    open var infiniteScrollBlock: ((InfiniteScrollView) -> Void)?
    
    // MARK: - RefreshPlugin
    open func isRefreshing(in scrollView: UIScrollView) -> Bool {
        return scrollView.fw_pullRefreshView?.state == .loading
    }
    
    open func shouldRefreshing(in scrollView: UIScrollView) -> Bool {
        return scrollView.fw_showPullRefresh
    }
    
    open func setShouldRefreshing(_ shouldRefreshing: Bool, in scrollView: UIScrollView) {
        scrollView.fw_showPullRefresh = shouldRefreshing
    }
    
    open func setRefreshing(block: @escaping () -> Void, customBlock: ((Any) -> Void)?, in scrollView: UIScrollView) {
        scrollView.fw_addPullRefresh(block: block)
        
        if let pullRefreshView = scrollView.fw_pullRefreshView {
            pullRefreshBlock?(pullRefreshView)
            customBlock?(pullRefreshView)
        }
    }
    
    open func setRefreshing(target: Any, action: Selector, customBlock: ((Any) -> Void)?, in scrollView: UIScrollView) {
        scrollView.fw_addPullRefresh(target: target, action: action)
        
        if let pullRefreshView = scrollView.fw_pullRefreshView {
            pullRefreshBlock?(pullRefreshView)
            customBlock?(pullRefreshView)
        }
    }
    
    open func beginRefreshing(in scrollView: UIScrollView) {
        scrollView.fw_triggerPullRefresh()
    }
    
    open func endRefreshing(in scrollView: UIScrollView) {
        scrollView.fw_pullRefreshView?.stopAnimating()
    }
    
    open func isLoading(in scrollView: UIScrollView) -> Bool {
        return scrollView.fw_infiniteScrollView?.state == .loading
    }
    
    open func shouldLoading(in scrollView: UIScrollView) -> Bool {
        return scrollView.fw_showInfiniteScroll
    }
    
    open func setShouldLoading(_ shouldLoading: Bool, in scrollView: UIScrollView) {
        scrollView.fw_showInfiniteScroll = shouldLoading
    }
    
    open func loadingFinished(in scrollView: UIScrollView) -> Bool {
        return scrollView.fw_infiniteScrollFinished
    }
    
    open func setLoadingFinished(_ loadingFinished: Bool, in scrollView: UIScrollView) {
        scrollView.fw_infiniteScrollFinished = loadingFinished
    }
    
    open func setLoading(block: @escaping () -> Void, customBlock: ((Any) -> Void)?, in scrollView: UIScrollView) {
        scrollView.fw_addInfiniteScroll(block: block)
        
        if let infiniteScrollView = scrollView.fw_infiniteScrollView {
            infiniteScrollView.showsFinishedView = showsFinishedView
            infiniteScrollBlock?(infiniteScrollView)
            customBlock?(infiniteScrollView)
        }
    }
    
    open func setLoading(target: Any, action: Selector, customBlock: ((Any) -> Void)?, in scrollView: UIScrollView) {
        scrollView.fw_addInfiniteScroll(target: target, action: action)
        
        if let infiniteScrollView = scrollView.fw_infiniteScrollView {
            infiniteScrollView.showsFinishedView = showsFinishedView
            infiniteScrollBlock?(infiniteScrollView)
            customBlock?(infiniteScrollView)
        }
    }
    
    open func beginLoading(in scrollView: UIScrollView) {
        scrollView.fw_triggerInfiniteScroll()
    }
    
    open func endLoading(in scrollView: UIScrollView) {
        scrollView.fw_infiniteScrollView?.stopAnimating()
    }
    
}
